### DISCUSSION TITLE
Closed Chain Robots

### DISCUSSION THREAD

**Discussion by sgmurray** [QuirkyCort/gears#96] [Ideas]
I think it would be cool to build robots like https://www.youtube.com/watch?v=9pjpQoZoW6E in the robot configurator. 
Currently, I don't think this is possible because there is a closed chain (it forms a loop) so it can't be represented as a tree.

This image explains the difference between open and closed chains:
https://slideplayer.com/slide/15307334/92/images/7/Open+and+Closed+Kinematic+Chains.jpg

One solution could be a "special hinge" component. The special hinge would have a option to specify a child component by name.

I know that it is possible to build open chain drawing robots e.g. https://www.youtube.com/watch?v=r8s6GkPP5EM . But, closed chains are used in other robots too, e.g. the lift mechanism on the  robot educator arm.

I am writing to get the GearsBot community's opinion. If you are a GearsBot user, would you be likely to make closed chain robots if the configurator supported them? If you are familiar with the existing GearsBot code, do you have any ideas on how to add support for closed chain robots to GEARS.

**Comment by QuirkyCort** [comment]
**Utility in the Robot**
I'm not sure about that. For my students trying to accomplish a robotics missions, I can't think of any cases where an open chain can't do just as well and with greater simplicity. In a real robot, the close chain has the advantage of not attaching a big and heavy actuator to the end of an arm, but that's not really applicable in a simulator.

**Utility in the World Builder**
There will be some utility here. Currently, some mechanism (eg. FLL missions) cannot be simulated correctly as the hinge are always open chain. Unlike in the robot, where close chain-like behaviors can be achieved by adding actuators to an open chain, this is not an option for world objects.

**Fun**
Could be a fun addition.

**Learning Value**
Could be of use to students learning kinematics, but that's college level and none of my students are ready for that.

**Adding support in Gears**
1) Would need to add in a unique identifier for each object. This is currently present in custom world (...not exposed to the user), but not in the robot.
2) The idea to specify a child component by name sounds good to me. It's not the most intuitive, but short of a major redesign of the GUI, I can't think of a better way.
3) Will also need to specify which end to connect to, as the actuator components have a base end and a moveable end.
4) Would need an additional pass to add the hinge, as it's possible that during the first pass, the child component may not have a physics body yet.
